### PR TITLE
Allow 'forced' rebuild of courses

### LINKED ISSUE #1497: Should be able to force a full rebuild of a course
When updating a plugin for an existing course it would be nice to have a force-rebuild button that rebuilds all the CSS + JS. At the moment I have to add a random extension and remove it everytime to make the Authoring Tool rebuild fully.

### PR DESCRIPTION
**Pull request by taylortom**
Holding ctrl/cmd changes the preview button to force a full build of the current course.

Fixes #1497.

### CODE REVIEW

File: frontend/src/core/sidebar/views/sidebarItemView.js [hunk 1]
```diff
@@ -32,6 +32,8 @@ define(function(require) {
         this.render();
         this.listenTo(Origin, 'sidebar:resetButtons', this.resetButtons);
         this.listenTo(Origin, 'sidebar:views:animateIn', this.animateViewIn);
+        this.listenTo(Origin, 'key:down', this.onKeyDown);
+        this.listenTo(Origin, 'key:up', this.onKeyUp);
```
Comment (tomgreenfield) [inline frontend/src/core/sidebar/views/sidebarItemView.js]: Missing `key:up` event trigger from origin.js in this branch.

File: frontend/src/core/sidebar/views/sidebarItemView.js [hunk 2]
```diff
@@ -94,7 +96,34 @@ define(function(require) {
...
+      window.clearTimeout(this.timeout);
+      this.timeout = -1;
+      switch(e.keyCode) {
+        case 17:
+        case 91:
+        case 93:
+          this.$('.editor-common-sidebar-preview')
+            .removeClass('action-primary')
+            .addClass('action-warning');
+          this.$('.editor-common-sidebar-preview-inner').text('Force rebuild');
```
Comment (tomgreenfield) [inline frontend/src/core/sidebar/views/sidebarItemView.js]: Could we localise this string please?

**Comment by taylortom** [comment]
@tomgreenfield have done as requested, and reworked a little.

File: frontend/src/core/sidebar/views/sidebarItemView.js [hunk 3]
```diff
@@ -100,22 +100,19 @@ define(function(require) {
...
-            .removeClass('action-primary')
-            .addClass('action-warning');
-          this.$('.editor-common-sidebar-preview-inner').text(window.polyglot.t('app.forcerebuild'));
-          this.forceRebuild = true;
-          this.timeout = window.setTimeout(_.bind(this.onKeyUp,this), 5000);
-          break;
-        default:
-          // do nothing
+      // 17: ctrl, 91: left cmd
+      if(e.keyCode !== 17 || e.keyCode !== 91) {
```
Comment (tomgreenfield) [inline frontend/src/core/sidebar/views/sidebarItemView.js]: Broken logic here – was this meant to be the below?
```js
if (!(e.keyCode === 17 || e.keyCode === 91))
```
I think the switch statement may read better.

Comment (taylortom) [inline frontend/src/core/sidebar/views/sidebarItemView.js]: Yep, the first one. Good spot!

**Comment by taylortom** [comment]
@tomgreenfield have amended

**Comment by taylortom** [comment]
I'll pick this one up again when we've merged the refactor

**Comment by taylortom** [comment]
@canstudios-louisem requested we look at putting this into v0.4.0, so have brought up-to-date with develop, and added the missing bits.

File: frontend/src/core/origin.js [hunk 4]
```diff
@@ -79,6 +79,7 @@ define(['require', 'underscore', 'backbone'], function(require, _, Backbone){
   // abstracted window events
   function listenToWindowEvents() {
     $(document).on('keydown', onKeyDown);
+    $(document).on('keyup', onKeyUp);
```
Comment (tomgreenfield) [inline frontend/src/core/origin.js]: While we're here could we cache selectors please?
```js
$(document).on({ keydown: onKeyDown, keyup: onKeyUp });
$(window).on({ resize: onResize, 'blur focus': onFocusBlur });
```

**Comment by taylortom** [comment]
@canstudios-louisem Tom rightly suggested that this is due to the native shortcuts for most browsers: ctrl+click usually opens a tab/new window in the background.

We may have to switch the key

**Comment by canstudios-louisem** [comment]
@taylortom I assume ctrl+m1 was chosen because its the same way you clear cache in chrome? Would we be better with a little dropdown or check box instead to force a rebuild? Nowhere else in the app uses keyboard shortcuts so it would also be better for user discoverability.

**Comment by taylortom** [comment]
Ok, have removed from this release, as I don't have time to rework this atm. We can maybe discuss more on the stand-up.

**Comment by lc-thomasberger** [comment]
We clould also use ctrl/cml + alt + click to bypass browsers actrl / shift click defaults.
Another option would be a bootrap-style split button --> https://getbootstrap.com/docs/4.0/components/dropdowns/#split-button-dropdowns

**Comment by taylortom** [comment]
Superceded by #2128

**Pull request closed without merging**